src/zenml/model/model_config.py

```diff
@@ -204,18 +204,19 @@ def _get_or_create_model_version(
         mv_request = ModelVersionRequestModel.parse_obj(
             self._get_request_params(ModelVersionRequestModel, model=model.id)
         )
-        mv = None
-        if self.recovery:
-            try:
-                mv = zenml_client.zen_store.get_model_version(
-                    model_name_or_id=self.name,
-                    model_version_name_or_id=self.version,
-                )
-            except KeyError:
+        try:
+            # trying to get existing `running` version
+            # if recovery is disable it can be only a version from current run
+            # if it exists - it was created in previous steps
+            mv = zenml_client.zen_store.get_model_version(
+                model_name_or_id=self.name,
+                model_version_name_or_id=self.version,
+            )
+        except KeyError:
+            if self.recovery:
```

Comment:
As the description of `recovery` states - it is not controlling that but rather controls if the `running` version will get deleted on failing or not. So this is by design.
`recovery: Whether to keep failed runs with new versions for later recovery from it.`

But I will rename it for better readability.

Reply:
My main point is: `create_new_model_version=True` will not always create new model versions. Which goes against what the docstring says:

> New version will be created if `create_new_model_version`, otherwise will try to fetch based on `version`.

Reply:
Make sense, I will reflect in docs better. Thanks!

Reply:
It will look like this after I commit:
```
        A new model is created implicitly if missing, otherwise existing model is fetched.

        A new model version is only created if `create_new_model_version` is True, but if there is an
        existing model version left over from the previous failed run with `delete_new_version_on_failure`
        is set to False, it will be fetched without creating a new one or deleting the old one.

        If `create_new_model_version` is set to False, a new model version will be fetched based on `version`.
```
Waiting for CI again :)

Reply:
- If there is an existing model version left over from the previous failed run with `delete_new_version_on_failure` set to False, it will be reused.
- Otherwise, if `create_new_model_version` is True, a new model version is created.
- If `create_new_model_version` is False and no version from a previous failed run exists, a model version will be fetched based on `version`.

Reply:
Maybe like this clearer?

Reply:
This is not accurate, cause the leftover version will only be used when a new version is requested, otherwise, if `version` is set - it will dominate and a specific version will be fetched.

My yet another attempt to make it clear and reflect reality:
```
        A new model is created implicitly if missing, otherwise existing model is fetched. Model
        name is controlled by the `name` parameter.

        Model Version returned by this method is resolved based on model configuration:
        - If there is an existing model version leftover from the previous failed run with
        `delete_new_version_on_failure` is set to False and `create_new_model_version` is True,
        leftover model version will be reused.
        - Otherwise if `create_new_model_version` is True, a new model version is created.
        - If `create_new_model_version` is False a model version will be fetched based on the version:
            - If `version` is not set, the latest model version will be fetched.
            - If `version` is set to a string, the model version with the matching version will be fetched.
            - If `version` is set to a `ModelStage`, the model version with the matching stage will be fetched.
```